aiokafka/client.py

```diff
@@ -1,489 +1,348 @@
-                        leader, None, None
-                    )
+            self.cluster.update_metadata(metadata)
+            break
+        else:
+            log.error('Unable to update metadata from %s', nodeids)
+            self.cluster.failed_update(None)
+            return False
+        return True
 
     @asyncio.coroutine
-    def send_metadata_request(self, payloads=()):
+    def fetch_all_metadata(self):
+        cluster_md = ClusterMetadata(
+            metadata_max_age_ms=self._metadata_max_age_ms)
+        metadata_request = MetadataRequest([])
+        nodeids = [b.nodeId for b in self.cluster.brokers()]
+        if 'bootstrap' in self._conns:
+            nodeids.append('bootstrap')
+        random.shuffle(nodeids)
+        for node_id in nodeids:
+            conn = yield from self._get_conn(node_id)
+            if conn is None:
+                continue
+            log.debug("Sending metadata request %s to %s",
+                      metadata_request, node_id)
 
-        encoder = KafkaProtocol.encode_metadata_request
-        decoder = KafkaProtocol.decode_metadata_response
+            try:
+                metadata = yield from conn.send(metadata_request)
+            except KafkaError as err:
+                log.error(
+                    'Unable to request metadata from node with id %s: %s',
+                    node_id, err)
+                continue
 
-        return (yield from self._send_broker_unaware_request(
-            payloads, encoder, decoder))
+            cluster_md.update_metadata(metadata)
+            break
+        else:
+            raise KafkaError('Unable to update metadata from %s', nodeids)
+        return cluster_md
 
-    def _build_resp(self, resps):
-        out = []
-        for resp in resps:
-            self._raise_on_response_error(resp)
-            out.append(resp)
-        return out
+    def add_topic(self, topic):
+        """Add a topic to the list of topics tracked via metadata.
 
-    @asyncio.coroutine
-    def send_produce_request(self, payloads=(), *, acks=1, ack_timeout=1.0):
+        Arguments:
+            topic (str): topic to track
         """
-        Encode and send some ProduceRequests
-
-        ProduceRequests will be grouped by (topic, partition) and then
-        sent to a specific broker. Output is a list of responses in the
-        same order as the list of payloads specified
+        if topic in self._topics:
+            return
+        self._topics.add(topic)
 
-        Params
-        ======
-        payloads: list of ProduceRequest
+    def set_topics(self, topics):
+        """Set specific topics to track for metadata.
 
-        Return
-        ======
-        list of ProduceResponse, in the order of input payloads
+        Arguments:
+            topics (list of str): topics to track
         """
+        if set(topics).difference(self._topics):
+            self._topics = set(topics)
+            return ensure_future(
```

Comment:
What logic behind this method? It returns awaitable or `None` so in one cases I can  use it with `yield from` in others not?